noir-r1cs/src/sparse_matrix.rs

```diff
@@ -175,10 +202,44 @@ impl Mul<HydratedSparseMatrix<'_>> for &[FieldElement] {
             rhs.matrix.num_rows,
             "Vector length does not match number of rows."
         );
+
+        let intermediate_multiplication =
+            LockFreeArray::new(vec![(0, FieldElement::zero()); rhs.matrix.num_entries()]);
+
+        let intermediate_reference = &intermediate_multiplication;
+
+        // Mapping phase
+        //
+        // Parallelize the multiplication
+        // Use a lock-free array to prevent constant resizing when collecting the
+        // iterator as the size is not known to Rayon. Collecting without a
+        // preallocating the intermediate vector is >15% slower
+        // Other options that have been explored
+        // - An IndexedParallelIterator on the values of the sparse matrix also wasn't
+        //   an option as it requires random access which we can't provide as we
+        //   wouldn't know the row a value belongs to. That's why the rows drive the
+        //   iterator below.
+        // - Acquiring a mutex per column in the result was too expensive (even with
+        //   parking_lot)
```

Comment:
What I had in mind was splitting the work over the mutable output vector, and creating a `iter_col` method on the matrix. (Which is substantially more complicated than `iter_row`, but should be doable).
Can we explore this option?